sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/StateBackedIterable.java

```diff
@@ -0,0 +1,223 @@
+ * limitations under the License.
+ */
+package org.apache.beam.fn.harness.state;
+
+import static org.apache.beam.runners.core.construction.ModelCoders.STATE_BACKED_ITERABLE_CODER_URN;
+
+import com.google.auto.service.AutoService;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.StateRequest;
+import org.apache.beam.runners.core.construction.CoderTranslation.TranslationContext;
+import org.apache.beam.runners.core.construction.CoderTranslator;
+import org.apache.beam.runners.core.construction.CoderTranslatorRegistrar;
+import org.apache.beam.sdk.coders.IterableLikeCoder;
+import org.apache.beam.sdk.fn.stream.DataStreams;
+import org.apache.beam.sdk.util.BufferedElementCountingOutputStream;
+import org.apache.beam.sdk.util.VarInt;
+import org.apache.beam.vendor.grpc.v1p26p0.com.google.protobuf.ByteString;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterables;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterators;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.io.ByteStreams;
+
+/**
+ * A {@link BeamFnStateClient state} backed iterable which allows for fetching elements over the
+ * portability state API. See <a
+ * href="https://s.apache.org/beam-fn-state-api-and-bundle-processing">remote references</a> for
+ * additional details.
+ *
+ * <p>One must supply a {@link StateBackedIterableTranslationContext} when using {@link
+ * CoderTranslator#fromComponents} to be able to create a {@link StateBackedIterable.Coder}.
+ */
+public class StateBackedIterable<T> implements Iterable<T> {
+
+  private final BeamFnStateClient beamFnStateClient;
+  private final org.apache.beam.sdk.coders.Coder<T> elemCoder;
+  @VisibleForTesting final StateRequest request;
+  @VisibleForTesting final List<T> prefix;
+
+  public StateBackedIterable(
+      BeamFnStateClient beamFnStateClient,
+      String instructionId,
+      ByteString runnerKey,
+      org.apache.beam.sdk.coders.Coder<T> elemCoder,
+      List<T> prefix) {
+    this.beamFnStateClient = beamFnStateClient;
+    this.elemCoder = elemCoder;
+
+    StateRequest.Builder requestBuilder = StateRequest.newBuilder();
+    requestBuilder
+        .setInstructionId(instructionId)
+        .getStateKeyBuilder()
+        .getRunnerBuilder()
+        .setKey(runnerKey);
+    this.request = requestBuilder.build();
+    this.prefix = prefix;
+  }
+
+  @Override
+  public Iterator<T> iterator() {
+    return Iterators.concat(
+        prefix.iterator(),
+        new DataStreams.DataStreamDecoder(
```

Comment:
I assume the DataStream here constructs iterator recursively right? so that not just *first* state chunk, but all the N state requests are handled.  

Just to double check, because the name "forFirstChunk" makes it a little confusing.

Maybe add brief comments here?

Reply:
Renamed to `readAllStartingFrom`